daemon/src/monitor.rs

```diff
@@ -279,30 +224,200 @@ where
+        self.current_status = new_status;
+
+        // 4. check for finished monitoring tasks
+        for ((txid, script), status) in self.current_status.iter() {
+            match self.awaiting_status.entry((*txid, script.clone())) {
+                Entry::Vacant(_) => {
+                    unreachable!("we are only fetching the status of scripts we are waiting for")
+                }
+                Entry::Occupied(mut occupied) => {
+                    let targets = occupied.insert(Vec::new());
+
+                    // Split vec into two lists, all the once for which we reached the target and
+                    // the once which we need to still monitor
+                    let (reached_monitoring_target, remaining) = targets
+                        .into_iter()
+                        .partition::<Vec<_>, _>(|(target_status, event)| {
+                            tracing::trace!(
+                                "{:?} requires {} and we have {}",
+                                event,
+                                target_status,
+                                status
+                            );
+
+                            status >= target_status
+                        });
+
+                    tracing::trace!("{} subscriptions reached their monitoring target, {} remaining for this script", reached_monitoring_target.len(), remaining.len());
+
+                    occupied.insert(remaining);
+
+                    for (target_status, event) in reached_monitoring_target {
+                        tracing::info!(%txid, target = %target_status, current = %status, "Bitcoin transaction reached monitoring target");
+                        self.cfd_actor_addr.send(event).await?;
+                    }
+                }
+            }
         }
 
         Ok(())
     }
 }
 
-pub struct StartMonitoring {
-    pub id: OrderId,
-    pub params: MonitorParams,
+fn print_status_change(txid: Txid, old: Option<&ScriptStatus>, new: &ScriptStatus) {
+    match (old, new) {
+        (None, new_status) if new_status > &ScriptStatus::Unseen => {
+            tracing::debug!(%txid, status = %new_status, "Found relevant Bitcoin transaction");
+        }
+        (Some(old_status), new_status) if old_status != new_status => {
+            tracing::debug!(%txid, %new_status, %old_status, "Bitcoin transaction status changed");
+        }
+        _ => {}
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
+struct Confirmed {
+    /// The depth of this transaction within the blockchain.
+    ///
+    /// Will be zero if the transaction is included in the latest block.
+    depth: u32,
+}
+
+impl Confirmed {
+    fn with_confirmations(blocks: u32) -> Self {
+        Self { depth: blocks - 1 }
+    }
+
+    /// Compute the depth of a transaction based on its inclusion height and the
+    /// latest known block.
+    ///
+    /// Our information about the latest block might be outdated. To avoid an
```

Comment:
thanks for the comment, I started thinking about the saturating sub again...

Reply:
This comment is a copy paste from the old code lol :D